pandas/core/frame.py

```diff
@@ -1909,41 +1909,73 @@ def to_dict(self, orient: str = "dict", into=dict):
             elif orient.startswith("i"):
                 orient = "index"
 
+        object_dtype_cols = {
+            col for col, dtype in self.dtypes.items() if is_object_dtype(dtype)
+        }
+        are_all_object_dtype_cols = len(object_dtype_cols) == len(self.dtypes)
```

Comment:
I think this might not work if the frame has duplicate column names, is that right? Should use indices instead.

Reply:
Hmm, appologies I hadn't considered the fact that one can use duplicacte column names, I think I have to completely use indices here, i.e. for `object_dtype_cols` also?

Reply:
Yes, that's correct.

Reply:
@rhshadrach changed to index based